### PR TITLE
Allow to print Junit report in human-readable format

### LINKED ISSUE #153: Allow to print Junit report in human-readable format
Hello,

Even, if a compact version (current behaviour of latest version 1.6.1) is acceptable when you use third-party tool, it's impossible for a human to get a look on it !

`console` format is then the only one we can read easily.

My Goal : Provides a JUnit report either in a compact version or a human-readable format.

I think it should be the original goal when we have a look on unit tests: 
https://github.com/shipmonk-rnd/composer-dependency-analyser/blob/1.6.1/tests/JunitFormatterTest.php#L79-L83
But in reality it's not TRUE !

I will propose a PR that fix this, and allow : 
- either to print a compact version (default behaviour) without `--verbose` option set
- a human-readable version when `--verbose` option is set

My second goal is to propose also a `checkstyle` format that is more readable and reusable in third-party tool context.

FYI: I've begun to work on a project https://github.com/llaville/sarif-php-converters that propose to project like your to provide a SARIF report (when they have at least a `checkstyle` output format).

I've already a prototype for a `CheckstyleFormatter` class.

Hope you'll like idea.

### PR DESCRIPTION
**Pull request by llaville**
Fixes #153

### CODE REVIEW

**Comment by llaville** [comment]
Introducing the new class `ShipMonk\ComposerDependencyAnalyser\Result\AbstractXmlFormatter` is in goal context to propose a `checkstyle` output format support

**Comment by JanTvrdik** [comment]
I don't think we need such option. We should just always pretty-print the output.

**Comment by llaville** [comment]
> I don't think we need such option. We should just always pretty-print the output.

I don't agree with you : my opinion is that 

- even if pretty print format has a ratio size to 20% greater than compact version, it's only used for debugging purpose by a human.
- when you use with a third-party tool/platform, machine-readable (compact) version is enough.

File: tests/JunitFormatterTest.php [hunk 1]
```diff
@@ -168,7 +169,7 @@ private function prettyPrintXml(string $inputXml): string
     {
         $dom = new DOMDocument();
         $dom->preserveWhiteSpace = false;
-        $dom->formatOutput = true;
+        $dom->formatOutput = true;  // always in human-readable format
```
Comment (janedbal) [inline tests/JunitFormatterTest.php:172]: I think we can now compare it 1:1 without altering the real output. It should just match the passed expected string.

File: tests/JunitFormatterTest.php [hunk 2]
```diff
@@ -177,9 +178,13 @@ private function prettyPrintXml(string $inputXml): string
         return trim($outputXml);
     }
 
+    /**
+     * @throws DOMException
```
Comment (janedbal) [inline tests/JunitFormatterTest.php:182]: Exceptions from other namespaces should never bubble up. Catch it at the lowest possible level and wrap to some our exception. That one can be propagated.

Doing that, you benefit from PHPStan exception analysis, see `checkedExceptionClasses`

File: src/Result/JunitFormatter.php [hunk 3]
```diff
@@ -117,12 +123,16 @@ public function format(
...
-            $xml .= $this->createUnusedIgnoresTestSuite($unusedIgnores);
+            $this->createUnusedIgnoresTestSuite($unusedIgnores);
         }
 
-        $xml .= '</testsuites>';
+        $xmlString = $this->document->saveXML(null, LIBXML_NOEMPTYTAG);
 
-        $this->printer->print($xml);
+        if ($xmlString === false) {
+            $xmlString = '';
```
Comment (janedbal) [inline src/Result/JunitFormatter.php:132]: Isnt this hiding some failure? I think this should be some exception.

Comment (janedbal) [reply src/Result/JunitFormatter.php:132]: Maybe you need something like [this](https://github.com/lightools/xml/blob/master/src/Xml/XmlLoader.php#L48) to get proper exception message.

Not sure tho...

File: src/Result/AbstractXmlFormatter.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * @var DOMDocument
+     */
+    protected $document;
+
+    /**
+     * @var DOMElement
+     */
+    protected $rootElement;
```
[the remark this answers is not in the document]
Comment (janedbal) [reply src/Result/AbstractXmlFormatter.php:25]: Same for `$document`.

File: src/Result/JunitFormatter.php [hunk 5]
```diff
@@ -170,38 +187,43 @@ private function createSymbolBasedTestSuite(string $title, array $errors, int $m
...
-        $xml .= '</testsuite>';
+        foreach ($errors as $packageName => $usagesPerClassname) {
+            $testcase = $this->document->createElement('testcase');
+            $testcase->setAttribute('name', $this->escape($packageName));
+            $testsuite->appendChild($testcase);
 
-        return $xml;
+            $failure = $this->document->createElement(
+                'failure',
+                $this->escape(implode('\n', $this->createUsages($usagesPerClassname, $maxShownUsages)))
```
[the remark this answers is not in the document]
Comment (janedbal) [reply src/Result/JunitFormatter.php:223]: Same above

**Comment by llaville** [comment]
Closing this PR because we do not have the same vision of reporting format and usage. 
@janedbal Thanks to spent time for code review

BTW, there are other enhancement that should be fixed first, to allow re-usability of your reports, otherwise no integration with third-party tools may be thinking .

**Comment by janedbal** [comment]
Ok

> there are other enhancement that should be fixed first, to allow re-usability of your reports

Anything else beside https://github.com/shipmonk-rnd/composer-dependency-analyser/issues/156 and things you already reported?

**Comment by llaville** [comment]
Yes and  https://github.com/shipmonk-rnd/composer-dependency-analyser/issues/157#issuecomment-2220470594

**Pull request closed without merging**